storm-core/src/jvm/backtype/storm/blobstore/BlobStoreAclHandler.java

```diff
@@ -0,0 +1,387 @@
+    boolean isSupervisor = false;
+    boolean isAdmin = false;
+    for(String u : user) {
+      if (_supervisors.contains(u)) {
+        isSupervisor = true;
+        break;
+      }
+      if (_admins.contains(u)) {
+        isAdmin = true;
+        break;
+      }
+    }
+    if (mask > 0 && !isAdmin) {
+      isSupervisor = (isSupervisor && (mask == 1));
+    }
+    return isSupervisor || isAdmin;
+  }
+
+  private boolean isNimbus(Subject who) {
+    Set<Principal> principals = null;
+    boolean isNimbusInstance = false;
+    if(who != null) {
+      principals = who.getPrincipals();
+      for (Principal principal : principals) {
+        if (principal instanceof NimbusPrincipal) {
+          isNimbusInstance = true;
+        }
+      }
+    }
+    return isNimbusInstance;
+  }
+
+  /**
+   * The user should be able to see the metadata if and only if they have any of READ, WRITE, or ADMIN
+   */
+  public void validateUserCanReadMeta(List<AccessControl> acl, Subject who, String key)
+      throws AuthorizationException {
+    Set<String> user = constructUserFromPrincipals(who);
+    if (isNimbus(who)) {
+      return;
+    }
+    if (isSupervisorOrAdmin(user, -1)) {
+      return;
+    }
+    for (AccessControl ac : acl) {
+      int allowed = getAllowed(ac, user);
+      LOG.debug(" user: {} allowed: {} key: {}", user, allowed, key);
+      if ((allowed & (READ | WRITE | ADMIN)) > 0) {
+        return;
+      }
+    }
+    throw new AuthorizationException(
+            user + " does not have access to " + key);
+  }
+
+  public void validateAnyACL(List<AccessControl> acl, int validPermissions, Subject who, String key) throws AuthorizationException {
+    Set<String> user = constructUserFromPrincipals(who);
+    LOG.debug("user {}", user);
+    if (isNimbus(who)) {
+      return;
+    }
+    if(isSupervisorOrAdmin(user, validPermissions)) {
+      return;
+    }
+    for (AccessControl ac : acl) {
+      int allowed = getAllowed(ac, user);
+      LOG.debug(" user: {} allowed: {} key: {}", user, allowed, key);
+      if ((allowed & validPermissions) > 0) {
+        return;
+      }
+    }
+    throw new AuthorizationException(
+            user + " does not have access to " + key);
+  }
+
+  //Here all acls must match
+  public void validateACL(List<AccessControl> acl, int mask, Subject who, String key)
+      throws AuthorizationException {
+    Set<String> user = constructUserFromPrincipals(who);
+    LOG.debug("user {}", user);
+    if (isNimbus(who)) {
+      return;
+    }
+    if(isSupervisorOrAdmin(user, mask)) {
+      return;
+    }
+    for (AccessControl ac : acl) {
+      int allowed = getAllowed(ac, user);
+      mask = ~allowed & mask;
+      LOG.debug(" user: {} allowed: {} disallowed: {} key: {}", user, allowed, mask, key);
+    }
+    if (mask == 0) {
+      return;
+    }
+    throw new AuthorizationException(
+            user + " does not have " + namedPerms(mask) + " access to " + key);
+  }
```

Comment:
@d2r Made the changes, hasPermissions and hasAnyPermissions for validateACL and validateAnyACL respectively and refactored the code.